File: DSA/Searching-Binary Search/binary_search.cpp
#include<bits/stdc++.h>
using namespace std;

int binarySearch(vector<int>arr,int target)
{
    int start = 0;
    int end = arr.size();

    while(start < end)
    {
        int mid = start  +(end - start)/2;
        
        if(arr[mid] == target)
            return mid;
        
        else if(arr[mid] > target)
        {
            // move to left part
            end = mid - 1;
        }
        else if(arr[mid] < target)
        {
            start = mid + 1;
        }
    }
    return -1;
}

// recursion 

int binarySearchRecursion(vector<int>&arr,int target,int start,int end)
{
    if(end > start){
        

    int mid = start + (end - start)/2;

    if(arr[mid] == target)
        return mid;
    
    else if(arr[mid] < target)
        return binarySearchRecursion(arr,target,mid + 1,end);
    
    return binarySearchRecursion(arr,target,start,mid - 1);
    }

    return -1;

}

int main()
{
    vector<int>arr = {1,8,15,17,25,27,29,35};
    int target = 15;
    // cout << binarySearch(arr,target);
    cout << binarySearchRecursion(arr,target,0,arr.size());
    return 0;
}

File: Python/Binary_Search/binary_Search.py
def binarySearch(arr,target):
    low = 0
    ans = -1
    high = len(arr) - 1
    while low <= high:
        mid = low + (high - low)//2
        
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:  # arr[mid] > target
            low = mid + 1
        else:
            high = mid - 1  
    return ans

arr = [1,3,5,5,5,12,16]
target = 5
# first= 2 last = 5
print(binarySearch(arr,target))